### PR TITLE
New Date Parser

### PR DESCRIPTION
**Pull request by cullophid**
The new parser accepts a format string and a date string, and parses the date based on the format.

The old parse function has be renamed to of

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -45,15 +45,51 @@ tomorrowAsString(new Date('2015-01-01')); // '2015-01-02';
 
 ## API
 
+### of
+`String | [Number] | Void -> Date`
```
Comment (TheLudd) [inline README.md]: If people want the current date they can use `new Date()`. I don't think that should be supported here.

Comment (cullophid) [inline README.md]: Thats a valid point

File: src/parse.js [hunk 2]
```diff
@@ -1 +1,96 @@
...
+  's': ['second', /\d{1,2}/, Number],
+  'SSS': ['millisecond', /\d{3}/, Number],
+  'SS': ['millisecond', /\d{2,3}/, Number],
+  'S': ['millisecond', /\d{1,3}/, Number],
+}
+const parseFormatString = format => format.match(DATE_TOKENS)
+
+// recursively chop off parts of the datestring and parse them based on the syntax object
+const parseDate = (result, datestring, syntax) => {
+  const touple = syntax[0]
```
Comment (svozza) [inline src/parse.js:42]: Nitpick: Should be `tuple`

Comment (cullophid) [inline src/parse.js:42]: Im a big fan of nitpicking
Naming was quite difficult here, so if you have any ideas for improvement Id like to hear them.
Does the code make sense ?

Comment (svozza) [inline src/parse.js:42]: I think `triplet` would probably be good here.

File: src/_spec/parse.js [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+
+    assert.equal(isValid(actual), false)
+  })
+
+  it('should return Invalid Date if given bad Date', () => {
+    const datestring = '2015-02-29'
+    const pattern = 'YYYY-MM-DD'
+    const actual = parse(pattern, datestring)
+
+    assert.equal(isValid(actual), false)
```
Comment (svozza) [inline src/_spec/parse.js:79]: Nice, I'm not a fan of parsers that turn this date into 1 March.

Comment (cullophid) [inline src/_spec/parse.js:79]: Yeah one of the motivations behind date-fp was to drop all the sillyness of js dates

Comment (svozza) [inline src/_spec/parse.js:79]: Agreed, an invalid date is an invalid date; you get into all sorts of messiness when you try to second guess what input you were provided.

**Comment by svozza** [comment]
This is looking good.

File: README.md [hunk 4]
```diff
@@ -46,14 +46,41 @@ tomorrowAsString(new Date('2015-01-01')); // '2015-01-02';
...
-const date = D.parse('2015-01-01') // Thu Jan 01 2015 00:00:00 GMT+0000 (GMT)
+const date = D.parse('YYYY-MM-DD HH:mm:ss', '2015-01-01 12:13:14') // Thu Jan 01 2015 12:13:14 GMT+0000 (GMT)
+const date = D.parse('MMM Do, YYYY', 'January 1st, 2015') // Thu Jan 01 2015 00:00:00 GMT+0000 (GMT)
 ```
 
+Parse supports the following tokens:
+
+|part             | Token |      Output         |
+|----------------:|:-----:|---------------------|
+|year             | YYYY  | 1971... 2015        |
```
Comment (svozza) [inline README.md:62]: Do we want some tests to test these year ranges or is that overkill?

Comment (cullophid) [inline README.md:62]: Im not entirely sure what the edge cases are...
but it doesn't hurt

Comment (svozza) [inline README.md:62]: Yeah, even just a case to see what what happens when you pass it 1970.

**Comment by cullophid** [comment]
Im gonna merge this in, If there are any issues with it lets open new PR / issues

**Comment by svozza** [comment]
Good stuff, although I still think `triplet` is a better variable name than `tuple`. ;)

**Comment by cullophid** [comment]
@svozza its not :)
